### ISSUE TITLE
[RPi] [Linux 5.10] [CIFS] Emby server fails to write to CIFS mount

### ISSUE THREAD

**Issue by MichaIng** [MediaBrowser/Emby#3676]
Since the recent Linux 5.10 kernel upgrade, Emby server fails to write to a CIFS mount with a denied permissions issue:
```
	System.UnauthorizedAccessException: System.UnauthorizedAccessException: Access to the path '/mount/infrastructure/emby_config/test/Emby Backup - 2021-02-17 23.10.0 - Auto/profile.txt' is denied.
	 ---> System.IO.IOException: Permission denied
	   --- End of inner exception stack trace ---
	   at System.IO.FileStream.WriteNative(ReadOnlySpan`1 source)
	   at System.IO.FileStream.FlushWriteBuffer()
	   at System.IO.FileStream.Dispose(Boolean disposing)
	   at System.IO.Stream.Close()
	   at System.IO.StreamWriter.CloseStreamFromDispose(Boolean disposing)
	   at System.IO.StreamWriter.Dispose(Boolean disposing)
	   at System.IO.TextWriter.Dispose()
	   at System.IO.File.WriteAllText(String path, String contents)
	   at Emby.Server.Implementations.IO.ManagedFileSystem.WriteAllText(String path, String text)
	   at MBBackup.ServerEntryPoint.ExecuteBackup(BackupProfile settings, CancellationToken cancellationToken, IProgress`1 progress, Boolean isAuto)
	   at MBBackup.Entities.ScheduledBackupTask.Execute(CancellationToken cancellationToken, IProgress`1 progress)
	   at Emby.Server.Implementations.ScheduledTasks.ScheduledTaskWorker.ExecuteInternal(TaskOptions options)
	Source: System.Private.CoreLib
	TargetSite: Void WriteNative(System.ReadOnlySpan`1[System.Byte])
	InnerException: System.IO.IOException: Permission denied
```
This implies backups, meta data (nfo) or previews/thumbnail media art. Empty files are created, but no content.

POSIX permissions are not the issue, even using emby:emby 777 does not work, only downgrading the kernel resolves the issue.

It has similarities with a known Mono/.NET core bug when it tries to create files on a file system that has no support for POSIX permissions. But that was fixed meanwhile, at least in .NET core, so I assume here as well that it is related to .NET, but I wonder how the kernel can play a role here.

To replicate:
- Install latest Raspberry Pi OS image on any Raspberry Pi model (Linux 5.4): https://downloads.raspberrypi.org/raspios_lite_armhf/images/raspios_lite_armhf-2021-01-12/2021-01-11-raspios-buster-armhf-lite.zip
- Install latest Emby server Debian package
- Mount any Samba share with `uid=emby`.
- Setup Samba share as media library and configure/use it to have Emby server creating/writing to a file inside the library directory.
- Repeat the same after upgrading the Raspberry Pi kernel (Linux 5.10): `apt install raspberrypi-kernel`

For reference: https://dietpi.com/phpbb/viewtopic.php?t=8686

**EDIT: It seems to be a general Linux 5.10 upstream issue, as with the Debian amd64 kernel it fails the same way.**

**Comment by LukePulverenti** [comment]
Hi, are you still running into this?

**Comment by MichaIng** [comment]
Yes same error with current Debian Bullseye kernel:
```
# uname -a
Linux VM-Bullseye 5.10.0-7-amd64 #1 SMP Debian 5.10.40-1 (2021-05-28) x86_64 GNU/Linux
```

Exactly the same setup on Debian Buster works:
```
# uname -a
Linux VM-Buster 4.19.0-17-amd64 #1 SMP Debian 4.19.194-1 (2021-06-10) x86_64 GNU/Linux
```

I moved the metadata directory to the CIFS mount, which has `0777` modes and `emby:emby` owner. As said above, it seems to be a change in the upstream Linux 5.10 CIFS kernel driver, as only upgrading the kernel broke it on Raspberry Pi.

**Comment by LukePulverenti** [comment]
Hi, I believe your issue is this:
https://emby.media/community/index.php?/topic/99596-data-deleted-from-media-folders-when-refreshing-all-metadata/&do=findComment&comment=1042461

In that thread I've linked to, you can find a workaround you can employ on your end. There is also a workaround we can add, and the next beta server will have this. Thanks.

**Comment by MichaIng** [comment]
The linked issue seems to be the same that I mentioned above, which we found with Mono software. Similarly, the mentioned workaround is to use the same user:group Emby runs as, for the Samba mount, so that this .NET function does not try to change POSIX permissions (which seems to be all the issue). But in our case, this does not help, and furthermore upgrading/downgrading the Linux kernel to/from 5.10 creates/solves the issue. It could be something in the CIFS driver as well.

Ah, now I see the `nobrl` option:
> Do not send byte range lock requests to the server. This is necessary for certain applications that break with cifs style mandatory byte range locks (and most cifs servers do not yet support requesting advisory byte range locks).

I'll try that one as well. Probably with Linux 5.10 (vs 5.4/4.19) the CIFS client starts to send POSIX advisory byte range lock requests instead of Windows style mandatory byte range locks, or so. There is another option which may be tried then:
> **forcemandatorylock**
> Even if the server supports posix (advisory) byte range locking, send only mandatory lock requests. For some (presumably rare) applications, originally coded for DOS/Windows, which require Windows style mandatory byte range locking, they may be able to take advantage of this option, forcing the cifs client to only send mandatory locks even if the cifs server would support posix advisory locks. forcemand is accepted as a shorter form of this mount option.

I'll play with those when I find time and looking forward for .NET Core 3.1 migration 🙂.

**Comment by LukePulverenti** [comment]
The 4.7 beta channel is running .NET 6 just FYI.

**Comment by LukePulverenti** [comment]
Resolved in the upcoming 4.7 server release (via updated .net runtime).
